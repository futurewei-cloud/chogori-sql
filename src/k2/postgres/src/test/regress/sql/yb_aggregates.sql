-- YB AGGREGATES TEST (for pushdown)

--
-- Test basic aggregates and verify overflow is handled properly.
--
CREATE TABLE ybaggtest (
    id         int PRIMARY KEY,
    int_2      int2,
    int_4      int4,
    int_8      int8,
    float_4    float4,
    float_8    float8
);

-- Insert maximum integer values multiple times to force overflow on SUM (both in DocDB and PG).
INSERT INTO ybaggtest VALUES (1, 32767, 2147483647, 9223372036854775807, 1.1, 2.2);
INSERT INTO ybaggtest
    SELECT series, t.int_2, t.int_4, t.int_8, t.float_4, t.float_8
    FROM ybaggtest as t CROSS JOIN generate_series(2, 100) as series;

-- Verify COUNT(...) returns proper value.
SELECT COUNT(*) FROM ybaggtest;
SELECT COUNT(0) FROM ybaggtest;
SELECT COUNT(NULL) FROM ybaggtest;

-- Delete row, verify COUNT(...) returns proper value.
DELETE FROM ybaggtest WHERE id = 100;
SELECT COUNT(*) FROM ybaggtest;
SELECT COUNT(0) FROM ybaggtest;

-- Verify selecting different aggs for same column works.
SELECT SUM(int_4), MAX(int_4), MIN(int_4), SUM(int_2), MAX(int_2), MIN(int_2) FROM ybaggtest;

-- Verify SUMs are correct for all fields and do not overflow.
SELECT SUM(int_2), SUM(int_4), SUM(int_8), SUM(float_4), SUM(float_8) FROM ybaggtest;

-- Verify shared aggregates work as expected.
SELECT SUM(int_4), SUM(int_4) + 1 FROM ybaggtest;

-- Verify NaN float values are respected by aggregates.
INSERT INTO ybaggtest (id, float_4, float_8) VALUES (101, 'NaN', 'NaN');
SELECT COUNT(float_4), SUM(float_4), MAX(float_4), MIN(float_4) FROM ybaggtest;
SELECT COUNT(float_8), SUM(float_8), MAX(float_8), MIN(float_8) FROM ybaggtest;

--
-- Test NULL rows are handled properly by COUNT.
--
-- Create table without primary key.
CREATE TABLE ybaggtest2 (
    a int
);

-- Insert NULL rows.
INSERT INTO ybaggtest2 VALUES (NULL), (NULL), (NULL);

-- Insert regular rows.
INSERT INTO ybaggtest2 VALUES (1), (2), (3);

-- Verify NULL rows are included in COUNT(*) but not in COUNT(row).
SELECT COUNT(*) FROM ybaggtest2;
SELECT COUNT(a) FROM ybaggtest2;
SELECT COUNT(*), COUNT(a) FROM ybaggtest2;

-- Verify MAX/MIN respect NULL values.
SELECT MAX(a), MIN(a) FROM ybaggtest2;

-- Verify SUM/MAX/MIN work as expected with constant arguments.
SELECT SUM(2), MAX(2), MIN(2) FROM ybaggtest2;
SELECT SUM(NULL::int), MAX(NULL), MIN(NULL) FROM ybaggtest2;

CREATE TABLE digit(k INT PRIMARY KEY, v TEXT NOT NULL);
INSERT INTO digit VALUES(1, 'one'), (2, 'two'), (3, 'three'), (4, 'four'), (5, 'five'), (6, 'six');
CREATE TABLE test(k INT PRIMARY KEY);
ALTER TABLE test ADD v1 int DEFAULT 5;
ALTER TABLE test ADD v2 int DEFAULT 10;
INSERT INTO test VALUES(1), (2), (3);
SELECT COUNT(*) FROM test;
SELECT COUNT(k) FROM test;
SELECT COUNT(v1) FROM test;
SELECT COUNT(v2) FROM test;
SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v2) AS count FROM test) AS c ON (d.k = c.count);
INSERT INTO test VALUES(4, NULL, 10), (5, 5, NULL), (6, 5, NULL);
SELECT COUNT(*) FROM test;
SELECT COUNT(k) FROM test;
SELECT COUNT(v1) FROM test;
SELECT COUNT(v2) FROM test;
SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(*) AS count FROM test) AS c ON (d.k = c.count);
SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(k) AS count FROM test) AS c ON (d.k = c.count);
SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v1) AS count FROM test) AS c ON (d.k = c.count);
SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v2) AS count FROM test) AS c ON (d.k = c.count);

DROP TABLE test;
DROP TABLE digit;

CREATE TABLE test(K INT PRIMARY KEY, v1 INT NOT NULL, v2 INT NOT NULL);
INSERT INTO test VALUES(1, 1, 1), (2, 2, 2), (3, 3, 3);
AlTER TABLE test DROP v1;
SELECT MIN(v2) FROM test;
SELECT MAX(v2) FROM test;
SELECT SUM(v2) FROM test;
SELECT COUNT(v2) FROM test;
