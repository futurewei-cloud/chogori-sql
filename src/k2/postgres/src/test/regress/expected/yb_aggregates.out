-- YB AGGREGATES TEST (for pushdown)
--
-- Test basic aggregates and verify overflow is handled properly.
--
CREATE TABLE ybaggtest (
    id         int PRIMARY KEY,
    int_2      int2,
    int_4      int4,
    int_8      int8,
    float_4    float4,
    float_8    float8
);
-- Insert maximum integer values multiple times to force overflow on SUM (both in DocDB and PG).
INSERT INTO ybaggtest VALUES (1, 32767, 2147483647, 9223372036854775807, 1.1, 2.2);
INSERT INTO ybaggtest
    SELECT series, t.int_2, t.int_4, t.int_8, t.float_4, t.float_8
    FROM ybaggtest as t CROSS JOIN generate_series(2, 100) as series;
-- Verify COUNT(...) returns proper value.
SELECT COUNT(*) FROM ybaggtest;
 count
-------
   100
(1 row)

SELECT COUNT(0) FROM ybaggtest;
 count
-------
   100
(1 row)

SELECT COUNT(NULL) FROM ybaggtest;
 count
-------
     0
(1 row)

-- Delete row, verify COUNT(...) returns proper value.
DELETE FROM ybaggtest WHERE id = 100;
SELECT COUNT(*) FROM ybaggtest;
 count
-------
    99
(1 row)

SELECT COUNT(0) FROM ybaggtest;
 count
-------
    99
(1 row)

-- Verify selecting different aggs for same column works.
SELECT SUM(int_4), MAX(int_4), MIN(int_4), SUM(int_2), MAX(int_2), MIN(int_2) FROM ybaggtest;
     sum      |    max     |    min     |   sum   |  max  |  min
--------------+------------+------------+---------+-------+-------
 212600881053 | 2147483647 | 2147483647 | 3243933 | 32767 | 32767
(1 row)

-- Verify SUMs are correct for all fields and do not overflow.
SELECT SUM(int_2), SUM(int_4), SUM(int_8), SUM(float_4), SUM(float_8) FROM ybaggtest;
   sum   |     sum      |          sum          |  sum  |  sum
---------+--------------+-----------------------+-------+-------
 3243933 | 212600881053 | 913113831648622804893 | 108.9 | 217.8
(1 row)

-- Verify shared aggregates work as expected.
SELECT SUM(int_4), SUM(int_4) + 1 FROM ybaggtest;
     sum      |   ?column?
--------------+--------------
 212600881053 | 212600881054
(1 row)

-- Verify NaN float values are respected by aggregates.
INSERT INTO ybaggtest (id, float_4, float_8) VALUES (101, 'NaN', 'NaN');
SELECT COUNT(float_4), SUM(float_4), MAX(float_4), MIN(float_4) FROM ybaggtest;
 count | sum | max | min
-------+-----+-----+-----
   100 | NaN | NaN | 1.1
(1 row)

SELECT COUNT(float_8), SUM(float_8), MAX(float_8), MIN(float_8) FROM ybaggtest;
 count | sum | max | min
-------+-----+-----+-----
   100 | NaN | NaN | 2.2
(1 row)

--
-- Test NULL rows are handled properly by COUNT.
--
-- Create table without primary key.
CREATE TABLE ybaggtest2 (
    a int
);
-- Insert NULL rows.
INSERT INTO ybaggtest2 VALUES (NULL), (NULL), (NULL);
-- Insert regular rows.
INSERT INTO ybaggtest2 VALUES (1), (2), (3);
-- Verify NULL rows are included in COUNT(*) but not in COUNT(row).
SELECT COUNT(*) FROM ybaggtest2;
 count
-------
     6
(1 row)

SELECT COUNT(a) FROM ybaggtest2;
 count
-------
     3
(1 row)

SELECT COUNT(*), COUNT(a) FROM ybaggtest2;
 count | count
-------+-------
     6 |     3
(1 row)

-- Verify MAX/MIN respect NULL values.
SELECT MAX(a), MIN(a) FROM ybaggtest2;
 max | min
-----+-----
   3 |   1
(1 row)

-- Verify SUM/MAX/MIN work as expected with constant arguments.
SELECT SUM(2), MAX(2), MIN(2) FROM ybaggtest2;
 sum | max | min
-----+-----+-----
  12 |   2 |   2
(1 row)

SELECT SUM(NULL::int), MAX(NULL), MIN(NULL) FROM ybaggtest2;
 sum | max | min
-----+-----+-----
     |     |
(1 row)

CREATE TABLE digit(k INT PRIMARY KEY, v TEXT NOT NULL);
INSERT INTO digit VALUES(1, 'one'), (2, 'two'), (3, 'three'), (4, 'four'), (5, 'five'), (6, 'six');
CREATE TABLE test(k INT PRIMARY KEY);
ALTER TABLE test ADD v1 int DEFAULT 5;
ALTER TABLE test ADD v2 int DEFAULT 10;
INSERT INTO test VALUES(1), (2), (3);
SELECT COUNT(*) FROM test;
 count
-------
     3
(1 row)

SELECT COUNT(k) FROM test;
 count
-------
     3
(1 row)

SELECT COUNT(v1) FROM test;
 count
-------
     3
(1 row)

SELECT COUNT(v2) FROM test;
 count
-------
     3
(1 row)

SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v2) AS count FROM test) AS c ON (d.k = c.count);
 k |   v   | count
---+-------+-------
 3 | three |     3
(1 row)

INSERT INTO test VALUES(4, NULL, 10), (5, 5, NULL), (6, 5, NULL);
SELECT COUNT(*) FROM test;
 count
-------
     6
(1 row)

SELECT COUNT(k) FROM test;
 count
-------
     6
(1 row)

SELECT COUNT(v1) FROM test;
 count
-------
     5
(1 row)

SELECT COUNT(v2) FROM test;
 count
-------
     4
(1 row)

SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(*) AS count FROM test) AS c ON (d.k = c.count);
 k |  v  | count
---+-----+-------
 6 | six |     6
(1 row)

SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(k) AS count FROM test) AS c ON (d.k = c.count);
 k |  v  | count
---+-----+-------
 6 | six |     6
(1 row)

SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v1) AS count FROM test) AS c ON (d.k = c.count);
 k |  v   | count
---+------+-------
 5 | five |     5
(1 row)

SELECT * FROM digit AS d INNER JOIN (SELECT COUNT(v2) AS count FROM test) AS c ON (d.k = c.count);
 k |  v   | count
---+------+-------
 4 | four |     4
(1 row)

DROP TABLE test;
DROP TABLE digit;
CREATE TABLE test(K INT PRIMARY KEY, v1 INT NOT NULL, v2 INT NOT NULL);
INSERT INTO test VALUES(1, 1, 1), (2, 2, 2), (3, 3, 3);
AlTER TABLE test DROP v1;
SELECT MIN(v2) FROM test;
 min
-----
   1
(1 row)

SELECT MAX(v2) FROM test;
 max
-----
   3
(1 row)

SELECT SUM(v2) FROM test;
 sum
-----
   6
(1 row)

SELECT COUNT(v2) FROM test;
 count
-------
     3
(1 row)
